plugins/outputs/otlp/client.go

```diff
@@ -0,0 +1,314 @@
+	"io/ioutil"
+	"net"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	metricsService "go.opentelemetry.io/proto/otlp/collector/metrics/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	// serviceConfig copied from OTel-Go.
+	// https://github.com/open-telemetry/opentelemetry-go/blob/5ed96e92446d2d58d131e0672da613a84c16af7a/exporters/otlp/grpcoptions.go#L37
+	serviceConfig = `{
+	"methodConfig":[{
+		"name":[
+			{ "service":"opentelemetry.proto.collector.metrics.v1.MetricsService" },
+			{ "service":"opentelemetry.proto.collector.trace.v1.TraceService" }
+		],
+		"retryPolicy":{
+			"MaxAttempts":5,
+			"InitialBackoff":"0.3s",
+			"MaxBackoff":"5s",
+			"BackoffMultiplier":2,
+			"RetryableStatusCodes":[
+				"UNAVAILABLE",
+				"CANCELLED",
+				"DEADLINE_EXCEEDED",
+				"RESOURCE_EXHAUSTED",
+				"ABORTED",
+				"OUT_OF_RANGE",
+				"UNAVAILABLE",
+				"DATA_LOSS"
+			]
+		}
+	}]
+}`
+)
+
+var (
+	maxErrorDetailStringLen = 512
+	maxTimeseriesPerRequest = 500
+)
+
+// Client allows reading and writing from/to a remote gRPC endpoint. The
+// implementation may hit a single backend, so the application should create a
+// number of these clients.
+type Client struct {
+	logger           log.Logger
+	url              *url.URL
+	timeout          time.Duration
+	rootCertificates []string
+	headers          metadata.MD
+	compressor       string
+
+	conn *grpc.ClientConn
+}
+
+// ClientConfig configures a Client.
+type ClientConfig struct {
+	Logger           log.Logger
+	URL              *url.URL
+	Timeout          time.Duration
+	RootCertificates []string
+	Headers          metadata.MD
+	Compressor       string
+}
+
+// NewClient creates a new Client.
+func NewClient(conf ClientConfig) *Client {
+	logger := conf.Logger
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	return &Client{
+		logger:           logger,
+		url:              conf.URL,
+		timeout:          conf.Timeout,
+		rootCertificates: conf.RootCertificates,
+		headers:          conf.Headers,
+		compressor:       conf.Compressor,
+	}
+}
+
+// getConnection will dial a new connection if one is not set.  When
+// dialing, this function uses its a new context and the same timeout
+// used for Store().
+func (c *Client) getConnection(ctx context.Context) (_ *grpc.ClientConn, retErr error) {
+	if c.conn != nil {
+		return c.conn, nil
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	useAuth := c.url.Scheme != "http"
+	_ = level.Debug(c.logger).Log(
+		"msg", "new OTLP connection",
+		"auth", useAuth,
+		"url", c.url.String(),
+		"timeout", c.timeout)
+
+	dopts := []grpc.DialOption{
+		grpc.WithBlock(), // Wait for the connection to be established before using it.
+		grpc.WithDefaultServiceConfig(serviceConfig),
+	}
+	if useAuth {
```

Comment:
I'd prefer to directly write
```suggestion
	if  c.url.Scheme != "http" {
```
to see what is going on here, but it's a matter of taste...

Reply:
After the refactor this was only used in one place, makes sense to remove the variable.